hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/replication/SimpleContainerDownloader.java

```diff
@@ -74,38 +74,28 @@ public SimpleContainerDownloader(
   public CompletableFuture<Path> getContainerDataFromReplicas(
       long containerId, List<DatanodeDetails> sourceDatanodes) {
 
-    CompletableFuture<Path> result = null;
-
     final List<DatanodeDetails> shuffledDatanodes =
         shuffleDatanodes(sourceDatanodes);
 
     for (DatanodeDetails datanode : shuffledDatanodes) {
       try {
-        if (result == null) {
-          result = downloadContainer(containerId, datanode);
-        } else {
-
-          result = result.exceptionally(t -> {
-            LOG.error("Error on replicating container: " + containerId, t);
-            try {
-              return downloadContainer(containerId, datanode).get();
-            } catch (ExecutionException | IOException e) {
-              LOG.error("Error on replicating container: " + containerId,
-                  e);
-            } catch (InterruptedException e) {
-              Thread.currentThread().interrupt();
-            }
-            return null;
-          });
-        }
+        CompletableFuture<Path> result =
+            downloadContainer(containerId, datanode);
+        result.get();
```

Comment:
The goal of this `getContainerDataFromReplicas` I think is to find a working source datanode, so it makes sense to call `get()` inside to wait for a successful retrieval (it's just getting the `Path` right? not the content of the container).

It looks like the only caller of this method `DownloadAndImportReplicator#replicate` is also handling the same set of exceptions when calling `CompletableFuture#get()`.

Is the `Path` retrieved twice if `get()` is called twice? If this is the case we probably should just straight return `result.get` (`Path`) for `getContainerDataFromReplicas` after the simplification? Could probably save a round trip here

CC @avijayanhwx  for more insights.

Reply:
I am pretty certain that calling `get()` on a future simply waits for the future to complete and returns its result (which is the path in this case). Calling `get()` while the download is still in progress just blocks until it completes or throws an exception. Calling `get()` again, I suspect just returns the result (path) again, but I will check to be sure.

Reply:
> It looks like the only caller of this method DownloadAndImportReplicator#replicate is also handling the same set of exceptions when calling CompletableFuture#get()

Yea I noticed that - perhaps I will change it to just return the path or null, and then we can remove the exception handling from the caller too.